QEP-label-path.rst

```diff
@@ -0,0 +1,71 @@
+.. _qep#[.#]:
+
+========================================================================
+QGIS Enhancement ??: Labeling path
+========================================================================
+
+:Date: October 2015
+:Author: Hugo Mercier
+:Contact: hugo dot mercier at oslandia dot com
+:Last Edited: 
+:Status:  
+:Version:
+
+Summary
+-------
+
+A new labeling option is proposed for addition here. It allows to use a user-defined linear path for the drawing of a label.
+
+Using a simplified path for the curved placement of labels is one of the possible use cases of such a feature.
+
+A user-defined path can be seen as an additional geometry on a layer, where the underlying model supports only one "geometry column".
+
+Proposed solution
+-----------------
+
+We propose the add a way for the user to use a different geometry object as support for the "curved" placement of a label.
+
+The current (2.11) version of QGIS allows to display labels along the linestring features in a "curved" manner.
+
+.. image:: curved_labeling.png
+
+However for linestrings defined with lots of control points (such as waterways) or with high curvatures, displaying correct
+labels can be either unesthetic or very slow.
+
+This is where using a path slightly different from the feature's geometry could be useful. For instance using a simplified version
+of the linestring could lead to a label faster to display or with smaller curvatures.
+
+We then propose to allow the "curved" placement to use an arbitrary geometry, through a data-defined "label path" property. The data-defined
+property could point to a user-defined column or to an expression.
```

Comment:
A core change which would be required to accompany this would be to allow access to other geometry columns defined within a layer. Eg for many data formats (Postgis, spatialite) it's possible to have more than one geometry column. These additional geometries are currently just converted to wkb representations and displayed as text fields by QGIS (consequently bogging down use of the attribute table for layers with complex secondary geometries). Support should be added so that additional geometry columns are stored as QgsGeometry objects, and then the label path could be directly bound to the secondary geometry column without the expense of going through the expression engine. This would also allow for the following use case:

- Polygon layer with a secondary line geometry attribute
- Data defined label path is bound to line geometry
- User can add the layer a second time to QGIS, using the line geometry as the main geometry column
- User can then edit the line geometries using the existing digitizing tools, and the labels for the polygon layer will instantly update to reflect any changes to paths.